src/mem/mem_scheduler.hh

```diff
@@ -0,0 +1,290 @@
+         * all of the flow control is hanled in this function.
+         *
+         * @param packet to send.
+         */
+        void sendPacket(PacketPtr pkt);
+
+        bool getBlocked(void);
+
+      protected:
+        /**
+         * Receive a timing response from the response port.
+         */
+        bool recvTimingResp(PacketPtr pkt) override;
+
+        /**
+         * Called by the response port if sendTimingReq was called on this
+         * request port (causing recvTimingReq to be called on the responder
+         * port) and was unsuccesful.
+         */
+        void recvReqRetry() override;
+
+        /**
+         * Called to receive an address range change from the peer responder
+         * port. The default implementation ignores the change and does
+         * nothing. Override this function in a derived class if the owner
+         * needs to be aware of the address ranges, e.g. in an
+         * interconnect component like a bus.
+         */
+        void recvRangeChange() override;
+    };
+
+    /**
+     * Handle the request from the CPU side
+     *
+     * @param requesting packet
+     * @return true if we can handle the request this cycle, false if the
+     *         requestor needs to retry later
+     */
+    bool handleRequest(CPUSidePort *port, PacketPtr pkt);
+
+    /**
+     * Handle the respone from the memory side
+     *
+     * @param responding packet
+     * @return true if we can handle the response this cycle, false if the
+     *         responder needs to retry later
+     */
+    bool handleResponse(PacketPtr pkt);
+
+    /**
+     * Handle a packet functionally. Update the data on a write and get the
+     * data on a read.
+     *
+     * @param packet to functionally handle
+     */
+    void handleFunctional(PacketPtr pkt);
+
+    /**
+     * Return the address ranges this memobj is responsible for. Just use the
+     * same as the next upper level of the hierarchy.
+     *
+     * @return the address ranges this memobj is responsible for
+     */
+    AddrRangeList getAddrRanges() const;
+
+    /**
+     * Tell the CPU side to ask for our memory ranges.
+     */
+    void sendRangeChange();
+
+    void wakeUp();
+
+    /// Instantiation of the CPU-side ports
+    std::vector<CPUSidePort*> cpuPorts;
+
+    /// Instantiation of the memory-side port
+    std::vector<MemSidePort*> memPorts;
+
+    void processNextReqEvent();
+    EventFunctionWrapper nextReqEvent;
+
+    void processNextRespEvent();
+    EventFunctionWrapper nextRespEvent;
+
+    MemSidePort* findMemoryPort(PacketPtr pkt);
+
+    const uint32_t readBufferSize;
+    const uint32_t writeBufferSize;
+    const uint32_t respBufferSize;
+    const uint32_t numberPorts;
+    const uint32_t numberQueues;
+    bool respBlocked;
+
+    // std::queue<PacketPtr> *readQueue;
+    // std::queue<PacketPtr> *writeQueue;
+    std::unordered_map<RequestorID, std::queue<PacketPtr> >::iterator currentReadEntry;
```

Comment:
Why needed?

Reply:
We use this as an arbiter over all the read queues in the memory scheduler